Userland/Libraries/LibVideo/VP9/Parser.cpp

```diff
@@ -1411,28 +1447,216 @@ DecoderErrorOr<i32> Parser::read_coef(Token token)
+            return;
+        }
+    }
+}
+
+void Parser::scale_mv(u8 ref_list, ReferenceFrame ref_frame)
+{
+    auto candidate_frame = m_candidate_frame[ref_list];
+    if (m_ref_frame_sign_bias[candidate_frame] != m_ref_frame_sign_bias[ref_frame])
+        m_candidate_mv[ref_list] *= -1;
+}
+
+void Parser::if_diff_ref_frame_add_mv(u32 candidate_row, u32 candidate_column, ReferenceFrame ref_frame, bool use_prev)
+{
+    for (auto ref_list = 0u; ref_list < 2; ref_list++)
+        get_block_mv(candidate_row, candidate_column, ref_list, use_prev);
+    auto mvs_are_same = m_candidate_mv[0] == m_candidate_mv[1];
+    if (m_candidate_frame[0] > ReferenceFrame::IntraFrame && m_candidate_frame[0] != ref_frame) {
+        scale_mv(0, ref_frame);
+        add_mv_ref_list(0);
+    }
+    if (m_candidate_frame[1] > ReferenceFrame::IntraFrame && m_candidate_frame[1] != ref_frame && !mvs_are_same) {
+        scale_mv(1, ref_frame);
+        add_mv_ref_list(1);
+    }
+}
+
+MotionVector Parser::clamp_mv(MotionVector vector, i32 border)
+{
+    i32 blocks_high = num_8x8_blocks_high_lookup[m_mi_size];
+    // Casts must be done here to prevent subtraction underflow from wrapping the values.
+    i32 mb_to_top_edge = -8 * (static_cast<i32>(m_mi_row) * MI_SIZE);
+    i32 mb_to_bottom_edge = 8 * ((static_cast<i32>(m_mi_rows) - blocks_high - static_cast<i32>(m_mi_row)) * MI_SIZE);
+
+    i32 blocks_wide = num_8x8_blocks_wide_lookup[m_mi_size];
+    i32 mb_to_left_edge = -8 * (static_cast<i32>(m_mi_col) * MI_SIZE);
+    i32 mb_to_right_edge = 8 * ((static_cast<i32>(m_mi_cols) - blocks_wide - static_cast<i32>(m_mi_col)) * MI_SIZE);
+
+    return {
+        clip_3(mb_to_top_edge - border, mb_to_bottom_edge + border, vector.row()),
+        clip_3(mb_to_left_edge - border, mb_to_right_edge + border, vector.column())
+    };
+}
+
+void Parser::clamp_mv_ref(u8 i)
+{
+    MotionVector& vector = m_ref_list_mv[i];
+    vector = clamp_mv(vector, MV_BORDER);
+}
+
+// 6.5.1 Find MV refs syntax
+void Parser::find_mv_refs(ReferenceFrame reference_frame, i32 block)
+{
+    m_ref_mv_count = 0;
+    bool different_ref_found = false;
+    u8 context_counter = 0;
+
+    m_ref_list_mv[0] = {};
+    m_ref_list_mv[1] = {};
+
+    MotionVector base_coordinates = MotionVector(m_mi_row, m_mi_col);
+
+    for (auto i = 0u; i < 2; i++) {
+        auto offset_vector = mv_ref_blocks[m_mi_size][i];
+        auto candidate = base_coordinates + offset_vector;
+
+        if (is_inside(candidate.row(), candidate.column())) {
+            auto candidate_index = get_image_index(candidate.row(), candidate.column());
+            auto index = get_image_index(candidate.row(), candidate.column());
+            different_ref_found = true;
+            context_counter += mode_2_counter[m_y_modes[index]];
+
+            for (auto ref_list = 0u; ref_list < 2; ref_list++) {
+                if (m_ref_frames[candidate_index][ref_list] == reference_frame) {
+                    // get_sub_block_mv()
```

Comment:
This is just to mark where the function defined by spec begins here. Since it's only called in one place and it was pointless to create a whole function call for it, I just wrote it in-line here.

I'll update the comment to be clearer.